File: Python/56-删除链表中重复的结点.py
# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteDuplication(self, pHead):
        # write code here
        new = ListNode(0)
        new.next = pHead
        pre, cur = new, pHead
        while cur:
            while cur.next and cur.next.val == cur.val:
                cur = cur.next
            if cur == pre.next:
                pre = pre.next
            else:
                pre.next = cur.next
            cur = cur.next
        return new.next

File: C++/56-删除链表中重复的结点.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution
{
public:
    ListNode *deleteDuplication(ListNode *pHead)
    {
        ListNode newHead = ListNode(0);
        newHead.next = pHead;
        ListNode *preHead = &newHead;
        ListNode *pre = preHead, *cur = pHead;
        while (cur)
        {
            while(cur->next && cur->next->val == cur->val)
                cur = cur->next;
            if(pre->next == cur)
                pre = pre->next;
            else
                pre->next = cur->next;
            cur = cur->next;   
        }
        return preHead->next;
    }
};